### ISSUE TITLE
2x decode regression vs mainline on gfx1151 + Mesa 25.2.8 (Vulkan/RADV, MTP speculative decoding, Qwen3.8-27B)

### ISSUE THREAD

**Issue by starbit-biostar** [halo-box/strix-llama.cpp#25]
### Summary

On a Strix Halo box (Ryzen AI MAX+ 395 / Radeon 8060S, gfx1151) this fork decodes at **~half the speed of mainline** for an MTP speculative-decoding workload on Vulkan/RADV, despite MTP being confirmed active in both builds. Reporting in case the mat-vec chunking dispatch is tuned against a different Mesa/RADV than ours.

### Environment

- Ryzen AI MAX+ 395, Radeon 8060S (RADV GFX1151), 128 GB unified LPDDR5X
- Ubuntu 24.04.4 LTS, kernel 7.0.0-30-generic
- Mesa **25.2.8-0ubuntu0.24.04.2** (RADV)
- Model: Qwen3.8-27B UD-Q4_K_XL (unsloth GGUF)

### Builds compared (identical CMake config)

| Build | Commit | cmake |
|---|---|---|
| mainline ggml-org/llama.cpp | d08c787 (2026-09-01) | `-DGGML_VULKAN=ON -DCMAKE_BUILD_TYPE=Release` |
| this fork | c7af5c6 (2026-09-05) | same |

### Repro

Same flags for both servers:

```
llama-server -m Qwen3.8-27B-UD-Q4_K_XL.gguf -ngl 999 -c 8192 --flash-attn on \
  --parallel 1 --spec-type draft-mtp --spec-draft-n-max 4
```

Workload: single-sequence chat completion, ~96-token prompt, ~228 generated tokens (translation-style, temp 0.1, thinking disabled), 3 runs each, idle GPU otherwise.

### Results (decode tok/s)

| Config | run 1 | run 2 | run 3 |
|---|---|---|---|
| mainline d08c787, static n-max 4 | 24.0 | 24.2 | 24.6 |
| fork c7af5c6, static n-max 4 | 13.1 | 13.1 | 13.1 |
| fork c7af5c6, `--spec-draft-adaptive` | 13.9 | 13.4 | 13.4 |

MTP is definitely engaged in the fork build — its logs show:

```
common_speculative_init_result: creating MTP draft context against the target model ...
slot print_timing: draft acceptance = 0.60274 (132 accepted / 219 generated), mean len = 3.20
```

so the ~2x gap looks like it's in the Vulkan compute path rather than the speculation logic. Non-speculative decode of this model on this box is ~10-13 tok/s, i.e. the fork's number is roughly what we get with speculation's gains erased.

### Question

Is the batched mat-vec chunking dispatch (the 3/5/6-column workaround mentioned in the README) calibrated on a specific Mesa/RADV version? On 25.2.8 it appears to regress rather than help. Happy to run patches or bisect builds on this hardware — the box is available for testing and we can turn benchmarks around quickly.

**Comment by LaurentZuijdwijk** [comment]
Thank you for submitting. I'll have a look

**Comment by julianmb** [comment]
Independent replication on a different Mesa, model, and n-max — same direction, smaller magnitude.

Strix Halo 128G, Mesa 26.1.7 (kisak), Qwen3.8-Flash-Next 125B-A6B (IQ4_XS-PLE 91G + Q8_0 MTP sidecar), n-max 6, same flags both sides. 8k MTP tg: ad914eb baseline 33.8/33.8/34.3 (n=3) → ea35c5066 29.1/29.1/29.0 (n=3). The triple is perfectly repeatable, so this isn't thermal noise — a clean −14%.

Two things that may narrow it down: (1) prefill on the same builds went *up* (+28–34% at 8k, +96% at 128k), so general matmul/compute got faster while MTP decode got slower — points at the draft/verify path rather than the chunking dispatch; (2) since this reproduces on Mesa 26.1.7 with a different model and n-max, it's unlikely to be 25.2.8-specific calibration. 128k MTP also regressed harder (swap cliff 0→3.2GB where the baseline runs clean).

Full receipts (logs + mem trajectories + method): julianmb/haloq38flash → results/MANIFEST.md, results/ea35c5-*.log. Same box/harness available if you want a specific bisect or patch tested — guarded runs, can turn them around same-day.

**Comment by LaurentZuijdwijk** [comment]
Thanks for the precise report. It reproduces here on the same GPU with Mesa 26.1.7, so it is not a driver-version calibration problem. Same model, same server flags, same kind of workload, decode tok/s over 3 runs:

| build | decode tok/s |
|---|---|
| mainline d08c787 | 19.7 / 19.7 / 20.2 |
| this fork c7af5c6 | 9.6 / 9.8 / 9.9 |
| this fork + `GGML_VK_MMV_NO_SPLIT=1` | 15.6 / 15.9 / 15.7 |

There are two causes, and they are independent:

1. **The mat-vec chunking is wrong for K-quants.** It was measured on Qwen3.8-Flash-Next (q8_0 / q6_K shapes) and chunks every K-quant batch into 2 columns without a measurement. On a dense Q4_K model the single dispatch scales fine at 3, 5, 6 and 8 columns, so chunking just re-reads the weights per chunk: llama-bench pp5 is 28.5 tok/s chunked against 43.1 unchunked. Fix: #27 restricts chunking to q8_0 and q6_K.

2. **MTP lost its recurrent-state rollback slots.** #18 removed the MTP type from `need_n_rs_seq()`, so the target context has `n_rs_seq = 0` and every partially accepted draft restores a checkpoint and re-decodes the accepted tokens. A verbose run shows 90 draft decodes for 48 accepted steps here, against 62 for 62 on mainline. Fix: #26, one line.

Each fix alone gets part of the way (12.1 and 15.0 tok/s); with both, decode is 19.9 / 20.0 / 20.0, i.e. mainline.

Until they merge, `GGML_VK_MMV_NO_SPLIT=1` gets you most of it on your box today. The open sync with the parent fork (#23) does not fix either; it only brings upstream's own Strix Halo mat-vec tuning, which is worth a few percent at 5 to 8 columns.

If you can run #26 + #27 on your Mesa 25.2.8 setup and post the numbers, that would confirm the driver version does not matter.

(Investigation and both PRs by an agent, Claude Fable 5.1, at the maintainer's request; measured on a Ryzen AI Max+ 395 / Radeon 8060S.)

**Comment by dabblingwithcode** [comment]
Reproduced on a third driver — Mesa 25.3.6 (RADV, gfx1151) — so the regression isn't specific to 25.2.8.

Setup: Ryzen AI Max+ 395 / Radeon 8060S (gfx1151), Fedora 43, Mesa 25.3.6.
Builds: halo-box c7af5c6 vs c7af5c6 + #26 + #27.
Model: unsloth Qwen3.8-27B UD-Q5_K_M (a K-quant, so it hits the same over-broad chunking as the UD-Q4_K_XL in the report). llama-bench -p 1,2,3,4,5,6,8 -n 0 -b 8 -r 3, idle GPU, pp tok/s:colsmaster c7af5c6 (chunk on)+#26+#27master + GGML_VK_MMV_NO_SPLIT=1318.2023.9330.60428.6540.3040.32524.2039.1449.86628.7646.1957.56831.8258.8168.83
The chunking regresses K-quant pp on 25.3.6 just as on 26.1.7/25.2.8 (col 8: 31.8 → 68.8 unchunked, ~2.2×).
#27 recovers most of it (col 8 → 58.8; matches NO_SPLIT exactly at col 4). The residual gap at cols 5/6/8 is this model's q6_K tensors, which #27 keeps chunked — consistent with your note, and it suggests q6_K chunking also regresses on sub-26.1 drivers, so the q6_K gate may want the same treatment.
Net: the regression tracks quant type, not driver version.(#27/chunking half only; #26 applied but MTP-decode A/B not yet run on this box).
Measured and written by Claude opus 5

**Comment by julianmb** [comment]
Independent data point from a second Strix Halo box (Ryzen AI MAX+ 395, 8060S, 128 GB, kernel 7.0.0, **Mesa 26.1.7-kisak** — note, newer than your 25.2.8), using ROCmFPX engine v1.7.0 on Vulkan0 with your exact repro shape (8192 ctx, FA on, `draft-mtp` n-max 4, temp 0.1):

| Config | Draft acceptance | Decode |
|---|---|---|
| Unsloth UD-Q4_K_M, no strict flag | **0.1%** (1/894, mean len 1.00) | ~7 tok/s |
| ROCmFP4_FAST, no strict flag | **0.1%** (1/894) | ~7 tok/s |
| ROCmFP4_FAST + `--spec-mtp-strict-qwen` | **4.2%** (17/402, mean len 1.17) | ~6 tok/s |

Two things that may help triage:

1. **This is a different failure mode from yours.** You get 60% acceptance with 2x slow decode (perf problem, consistent with your mat-vec chunking hypothesis). I get ~0% acceptance (engagement problem) — MTP never gets going on Vulkan0 in this engine at all, while the same engine on ROCm0 shows the published 2.4–2.9x MTP wins. Worth checking whether your fork has the same split: does MTP engage equally on both backends?
2. **The UD-K_M file is fine** — it carries 5814 MTP tensor refs, so the 0% is not a missing-heads issue. (Caveat: I ran K_M, you ran XL — if XL-vs-K_M matters, I can pull XL and re-run.)

Box available for patch testing / bisect if useful — scripted harness, can turn numbers around same-day.

**Comment by LaurentZuijdwijk** [comment]
@julianmb good to see you here. If you have time could you test the 2 PRs posted above?

**Comment by julianmb** [comment]
Tested both PRs on Flash-Next 125B (IQ4_XS-PLE 91G + Q8_0 sidecar, n-max 6, Mesa 26.1.7, server-based 8k cells with draft stats):

| engine | pp | tg | drafts (acc) |
|---|---|---|---|
| ad914eb (control) | 392.3 | 31.5 | 29/25 (86%) |
| halo-box c7af5c6 | 427.1 | 25.65 | 36/21 (58%) |
| + #26 | 429.9 | 26.04 | 36/21 (identical) |
| + #26 + #27 | 428.7 | 25.95 | 36/21 (identical) |

Neither PR moves this workload (all within noise, draft sequences identical across the three halo-box runs) — consistent with #27 being K-quant-specific.

But there's a second gap worth knowing: halo-box trails ad914eb by −19% tg on Flash-Next with much lower acceptance (58% vs 86%; 36 vs 29 draft rounds for the same prompt, temp 0). Prefill is *higher* on halo-box, so it's the same prefill-up/decode-down split as a regression I measured on Nathan's branch (33.8→29.1). Hypothesis: newer kernels shifted draft-path numerics enough to flip greedy draft tokens → lower acceptance → more rounds. Your two PRs don't touch it.

Receipts: julianmb/haloq38flash → results/halobox-*.log/.json + MANIFEST. Happy to run follow-ups (logit comparison, other n-max) on this box.

**Comment by voidsurfer** [comment]
Independent two-backend check of #26 + #27, on a third box, against post-sync master.

Setup: Ryzen AI Max+ 395, Radeon 8060S (gfx1151), Mesa 26.1.5 (RADV), Vulkan 1.4.354. Base is master 88c87dd6. This is newer than the c7af5c6 base used above: it includes the 2026-09-06 upstream sync (#23). Both PR heads apply to it without conflicts. Test arm is 88c87dd6 + #26 (c69f0e8c) + #27 (d22fa655). The full diff is 25 lines.

Model note: we do not have the unsloth UD-Q4_K_XL file. We used the Huihui abliterated fine-tune of the same recipe. The tensor mix is the same class: Q4_K and Q5_K dominant, plus 56 q6_K and 110 q8_0 tensors. The native MTP head is in the file (blk.64.nextn layer, shared output head). Acceptance behavior of this fine-tune is its own, so we compare arms only and do not compare our absolute tok/s with the numbers above.

Workload: the issue's shape. Single-slot server, `-ngl 999 -c 8192 --flash-attn on --parallel 1 --cache-type-k f16 --cache-type-v f16 --spec-type draft-mtp --spec-draft-n-max 4 -b 1024 -ub 512 -t 16`. Prompt is a 107-token template render with `enable_thinking=false`, translation style. temp 0.1, seed 42, `cache_prompt` off. Five runs per arm, arm order rotated, fresh server per run. Each server start was checked against the expected build banner, model path and n_rs_seq value before any request ran.

Decode tok/s, mean +/- SD, n=5:

| backend | master 88c87dd6 | + #26 + #27 | delta |
|---|---:|---:|---:|
| Vulkan (RADV) | 11.889 +/- 0.072 | 23.177 +/- 0.071 | +94.9% |
| HIP (ROCm 10.0) | 15.289 +/- 0.114 | 21.751 +/- 0.097 | +42.3% |

Mechanism, both backends: n_rs_seq 0 -> 4; recurrent buffer 149.62 -> 748.12 MiB (+598.5 MiB on device); speculative checkpoints per run 21 -> 0; checkpoint restores 14 -> 0. Token streams are identical across the five repeats within each arm. Acceptance ratio and mean accepted length differ between arms (0.763 vs 0.630 accepted per slot; 2.25 vs 2.52 per event). The two paths verify different slot counts, so we report these two metrics separately and draw no conclusion from the difference alone.

HIP detail: same two pins, built and run in a container (ROCm 10.0, gfx1151). The focused MUL_MAT suite has zero failing cases on both arms on HIP. The three cases that still fail on Vulkan with #27 (bf16 n=9, bf16 n=16, q8_0 n=4 at m=128, k=5120) pass on HIP on both arms. We did not run with the ring buffer PR. Treat the HIP column as data taken without that prerequisite, not as a full ROCm validation.

One correctness observation. The MTP output diverges between master and the fixed arm at exactly token 56 (master 45756, fixed arm 15562), with the same token IDs on both backends. #27 is inert on HIP, so this is not the chunking dispatch. The candidate cause is the rollback path itself. This matches the property named in Nathan's fork when it moved MTP to full checkpoints: the RS-plane rollback "can diverge slightly after rejected drafts" (Nathanw1014/llama.cpp, revert a17e8432b; his current state is f25eefeaf, full checkpoints with a livelock fix). Our data shows this divergence is real and backend-independent. We did not compare against a ggml-org build, so we do not call it a bug. Separately, plain non-speculative decode diverges between arms on Vulkan at token 15 only; `GGML_VK_MMV_NO_SPLIT=1` on master reproduces the fixed arm's stream exactly, so that one is the dispatch, not #26.

Summary: #26 + #27 give a large measured MTP decode win on both backends for this model class, with no new failing backend cases and byte-identical prefill logits on four fixed corpora. #26 trades exactness for speed: it restores the upstream RS rollback, which can change tokens after rejected drafts. Nathan's fork keeps token-exact rollback with full checkpoints and pays the restore cost that this issue reports. We think one of the two should be picked with that trade known and stated.

Raw logs, commands, hashes and per-run summaries are local; we can attach any part on request.

Assisted-by: Codex (runs and checks done locally on real hardware by the poster).

-Astra

**Comment by julianmb** [comment]
Follow-up with byte-level outputs (same runs, full completion JSONs):

- base == +#26 == +#26+#27, **byte-identical** across all three 128-token completions. So on Flash-Next/n-max-6, #26 changes neither speed (±1.5%) nor a single output token — fully inert here. That bounds voidsurfer's exactness question for this workload: no divergence cost *and* no gain.
- ad914eb vs halo-box-base diverge at **char 29** on the same prompt ("very chatty mood" vs "really enthusiastic today"), behind different draft trajectories (29 vs 36 drafts). So the 86%→58% acceptance gap is engine-level greedy divergence that predates both PRs — not something #26/#27 introduce or fix.

Net: the two PRs are validated harmless on this model (adds to dabblingwithcode's #27 half), but the Flash-Next gap vs ad914eb lives elsewhere — likely draft-path numerics, per my earlier note. Same receipt paths as before.

**Comment by starbit-biostar** [comment]
Confirmed on the original report's setup — **Mesa 25.2.8** (RADV gfx1151, Ryzen AI MAX+ 395 / Radeon 8060S, kernel 7.0.0-30): #26 + #27 recover mainline parity here as well, so driver version indeed does not matter.

Same protocol as the original report (Qwen3.8-27B UD-Q4_K_XL, `-ngl 999 -c 8192 --flash-attn on --parallel 1 --spec-type draft-mtp --spec-draft-n-max 4`, identical 3-prompt translation workload, idle GPU, 3 runs):

| build | decode tok/s |
|---|---|
| mainline d08c787 | 24.0 / 24.2 / 24.6 |
| fork c7af5c6 | 13.1 / 13.1 / 13.1 |
| fork c7af5c6 + #26 + #27 | **23.2 / 23.3 / 23.2** |

Draft acceptance is unchanged by the patches (0.575, mean accepted length 3.30 — vs 0.60 / 3.2 measured on the unpatched fork), which is consistent with both fixes being pure execution-path issues rather than anything in drafting itself.

Thanks for the fast turnaround on the diagnosis — happy to re-test once #26/#27 merge.

(Benchmarks run by an agent, Claude Fable 5, on our hardware.)

**Comment by starbit-biostar** [comment]
Post-merge confirmation on the original setup (Mesa 25.2.8, RADV gfx1151, Ryzen AI MAX+ 395): master `7449a0fe9` decodes at 26.1-26.3 tok/s on the same protocol (Qwen3.8-27B UD-Q4_K_XL, 8k ctx, FA on, draft-mtp n-max 4, temp 0.1) - above both the pre-merge #26+#27 test (23.2-23.3) and mainline `d08c787` (~24), with draft acceptance up to 0.67 (mean len 3.7). Regression fully resolved here; thanks for the quick fixes. (Benchmarks run by an agent, Claude Fable 5, on our hardware.)
